core/tx_list.go

```diff
@@ -529,21 +555,36 @@ func (l *txPricedList) Underpriced(tx *types.Transaction) bool {
 // Note local transaction won't be considered for eviction.
 func (l *txPricedList) Discard(slots int, force bool) (types.Transactions, bool) {
 	drop := make(types.Transactions, 0, slots) // Remote underpriced transactions to drop
-	for len(*l.remotes) > 0 && slots > 0 {
-		// Discard stale transactions if found during cleanup
-		tx := heap.Pop(l.remotes).(*types.Transaction)
-		if l.all.GetRemote(tx.Hash()) == nil { // Removed or migrated
-			l.stales--
-			continue
+	for slots > 0 {
+		if len(l.urgent.list)*floatingRatio > len(l.floating.list)*urgentRatio || floatingRatio == 0 {
```

Comment:
I don't think it's a good idea to rebalance things if they are not full. We should keep shoving txs into the urgent queue until it fills up and only then overflow into the floating one. Am I missing something perhaps? I don't see any upside, only downsides to moving txs into the floating queue while there's room in the urgent queue.

Reply:
In `txPricedList` the definition of full is that items are being discarded. The `urgentRatio : floatingRatio` thing is internal. If I wanted to enforce absolute limits here then I would need to deal with slot counting because that's how the global limit is calculated in the tx pool. I don't think that complexity is necessary. I think this is a clearly defined behavior: if we discard for any reason at any point, we discard from the two queues alternating in a 4:1 ratio (before which a rebalancing is automatically performed so that everyone is where they should be). We don't assume anything very specific about the caller but this is a safe and rational strategy for any meaningful use of `txPricedList`.